tests/FakeItEasy.Specs/WrappingFakeSpecs.cs

```diff
@@ -272,6 +274,28 @@ public static void NotAWrappingFakeCallsWrappedMethod(IFoo fake, Exception excep
                     .WithMessage("The configured fake is not a wrapping fake."));
         }
 
+        [Scenario]
+        public static void EventSubscriptionSubscribesToWrappedEvent(Foo realObject, IFoo wrapper, EventHandler handler)
+        {
+            "Given a real object"
+                .x(() => realObject = new Foo());
+
+            "And a fake wrapping this object"
+                .x(() => wrapper = A.Fake<IFoo>(o => o.Wrapping(realObject)));
+
+            "And an event handler"
+                .x(() => handler = A.Fake<EventHandler>());
+
+            "When the handler is subscribed to the fake's event"
+                .x(() => wrapper.SomeEvent += handler);
+
+            "And the event is raised"
+                .x(() => realObject.RaiseSomeEvent());
+
+            "Then the handler is called"
+                .x(() => A.CallTo(handler).MustHaveHappened());
+        }
+

```

Comment:
This should cover the scenario described by @mlaily.
We might need to check other scenarios as well, but I'm not sure which...

Reply:
I wouldn't mind seeing what happens when one uses `+= Raise…` to raise the event in this case. I don't expect it will work, and that could be fine, but I think it's worth trying, to see the effect. Heck, I might even keep the test with a comment indicating that we expect the behaviour.

I'd also check the interaction of event raising with `CallsBaseMethods`. And strict? I expect previously we could subscribe to events on a strict fake (and raise them). We won't be able to now. Which I think is fine?

I find I still don't have a complete handle on the interaction between these features. I'm going to need to think for a big before we merge, I think.

Reply:
Indeed, `+= Raise...` won't work. And I don't see how we could make it work, since we have no control on how the wrapped raises the event. Unless... we could do something like this:
- keep the default behavior that supports the `+= Raise...` syntax
- *also* forward the event subscription to the wrapped object
- when the wrapped object raises the event, the registered handlers are called directly by the wrapped object
- when `+= Raise` is used, the registered handlers are called by `EventRule`
This way, we still support both "modes". Of course, this approaches has its own problems. E.g. what if the wrapped object decides to clear the handlers? We wouldn't know about it and the handlers managed by `EventRule` would no longer be synchronized with the wrapped object. Also, the `sender` argument would be different depending on how the event was raised (maybe not a big issue, since the user can specify the sender in `Raise.With`)

> I'd also check the interaction of event raising with CallsBaseMethods.

What do you mean? `CallsBaseMethods` is not related to wrapping fakes, as far as I can tell.

> And strict? I expect previously we could subscribe to events on a strict fake (and raise them). We won't be able to now. Which I think is fine?

I think this is the correct behavior. I can add a spec to cover this.

Reply:
> If nothing else, if we can figure out a way to relax a strict fake so that it could start handling and raising events

Would it make sense, though? Everything would be strict except events? If we did that, why wouldn't we also add similar relaxations for properties or methods? It seems a little strange to do that just for events (unless you just want a way to restore the previous behavior, for compability?).

> in case there's a client out there who'd been making strict fakes and doing nothing but raising events.

Before this PR, `Strict` didn't affect event raising and subscription, so there was no point in making the fake strict. If anything, the existing behavior was a bug.

Reply:
> True, but it's the interaction between the event rule and calling the base methods that I worry about

Ah, I see. I guess that currently, base methods aren't called for event subscription, so this would change with this PR

Reply:
> I wouldn't mind seeing what happens when one uses += Raise… to raise the event in this case.

I added a spec for this case. As expected, using `+= Raise.With...` doesn't raise the event. But it does forwards the "dummy" handler to the wrapped object... So when the wrapped object actually raises the event, it throws an exception, since this handler is not actually supposed to be called. We probably need to do something to prevent that. I can see two ways to do it:
1. Split `EventRule` into two rules
  - One that handles the `+= Raise.With...` syntax (pre-user rule)
  - One that handles event subscription (post-user rule)
For a wrapping fake, `Raise.With` wouldn't actually do anything, since there would be no handlers registered on `EventRule`.
2. Handle it in `WrappedObjectRule` and throw an exception if the user tries to use `Raise.With`

I lean toward the second option, because it gives clearer feedback to the user

Reply:
> Handle it in `WrappedObjectRule` and throw an exception if the user tries to use `Raise.With`

I also prefer this option, as the first one sounds like we'de be creating a variant of #1795.

I think we'd need a similar loophole for a Fake that `CallsBaseMethods`?

Reply:
> > If nothing else, if we can figure out a way to relax a strict fake so that it could start handling and raising events
> 
> Would it make sense, though? Everything would be strict except events? If we did that, why wouldn't we also add similar relaxations for properties or methods? It seems a little strange to do that just for events (unless you just want a way to restore the previous behavior, for compability?).
> 
> > in case there's a client out there who'd been making strict fakes and doing nothing but raising events.
> 
> Before this PR, `Strict` didn't affect event raising and subscription, so there was no point in making the fake strict. If anything, the existing behavior was a bug.

I agree that it was a bug, but one that was very long-standing, so people may be relying on it. I was a little unclear. I wasn't suggesting we have a "Strict except events" mode. Just that we figure out how to allow clients to support event-registration and -raising on an otherwise strict Fake. It would help if we get a question. Maybe someone out there is using implicit creation options to make every Fake strict and they have some even-only interface. I know, it's not super likely.

But the main reason I mention it is that I know how to work around autoproperties on strict Fakes. We just make a setter that configures the getter to return the set value. I have _no idea_ how to turn on raising of events via `+=`. It would be nice to know if it's even possible to restore the functionality we're potentially taking away.

Reply:
> I think we'd need a similar loophole for a Fake that `CallsBaseMethods`?

Do you mean `CallsWrappedMethod` ?
Yes. Unless we just let the exception be thrown when Raise.Now is called... But [the message is not great](https://github.com/FakeItEasy/FakeItEasy/blob/5b73e426467a83de6e65cc3a4337e98c57379c8a/src/FakeItEasy/ExceptionMessages.cs#L16), and anyway it's probably better to throw as early as possible.

Reply:
> I wasn't suggesting we have a "Strict except events" mode. Just that we figure out how to allow clients to support event-registration and -raising on an otherwise strict Fake.

Ah, I see. To be honest, I'm not sure I like the idea. I'd rather make it a breaking change and make strict fakes really strict.

> But the main reason I mention it is that I know how to work around autoproperties on strict Fakes. We just make a setter that configures the getter to return the set value. I have no idea how to turn on raising of events via +=. It would be nice to know if it's even possible to restore the functionality we're potentially taking away.

Yes, I was thinking about that too. I don't think something like `A.CallTo(() =fake.Event += A<EventHandler>._)` is currently supported... And it probably couldn't be supported, because this would be legal in an expression anyway. We could do this instead:

```csharp
    EventHandler handler = null;
    A.CallTo(fake).Where(call => call.Method.Name == "add_Event").Invokes((EventHandler h) => handler += h);
    A.CallTo(fake).Where(call => call.Method.Name == "remove_Event").Invokes((EventHandler h) => handler -= h);
```
We might want to introduce an easier syntax, though...

Reply:
Maybe `A.CallToSubscribe(fake, "Event")` instead of `A.CallTo(fake).Where(call => call.Method.Name == "add_Event")`?

Reply:
>> I wasn't suggesting we have a "Strict except events" mode. Just that we figure out how to allow clients to support event-registration and -raising on an otherwise strict Fake.
>
>Ah, I see. To be honest, I'm not sure I like the idea. I'd rather make it a breaking change and make strict fakes really strict.

I can see I still wasn't clear. I would prefer strict Fakes to be strict as well. I just thought we should understand the technique (if there can be one) for adapting a Fake so that if one wanted to add the event-handling/raising behaviour after the fact, one could, just like one can explicitly configure other methods on a strict Fake.

> We could do this instead:
> 
> ```c#
>    EventHandler handler = null;
>    A.CallTo(fake).Where(call => call.Method.Name == "add_Event").Invokes((EventHandler h) => handler += h);
>    A.CallTo(fake).Where(call => call.Method.Name == "remove_Event").Invokes((EventHandler h) => handler -= h)
> ```

This is the sort of thing I was thinking of figuring out, yes. Then we could just raise by calling `handler`. Fantastic.
I'm not in a huge rush to find syntactic sugar, but if we did look for some, I think maybe something that does a little more work (since I think your example would still require the user to write the `Invokes` and so on…)?

>> I think we'd need a similar loophole for a Fake that `CallsBaseMethods`?
>
> Do you mean `CallsWrappedMethod`?

No, I meant that a Fake created with the `CallsBaseMethods` option would probably also need special handling, just as ones created that wrap a concrete object would.

Reply:
> I can see I still wasn't clear. I would prefer strict Fakes to be strict as well. I just thought we should understand the technique (if there can be one) for adapting a Fake so that if one wanted to add the event-handling/raising behaviour after the fact, one could, just like one can explicitly configure other methods on a strict Fake.

OK, I think I understand this time.

> since I think your example would still require the user to write the Invokes

Indeed.

> No, I meant that a Fake created with the CallsBaseMethods option would probably also need special handling, just as ones created that wrap a concrete object would.

Oh. Sorry, I'm a bit slow today... Yes, that also applies to CallsBaseMethods.